addons/promote/common/src/main/java/org/commonjava/indy/promote/data/PromotionManager.java

```diff
@@ -761,430 +561,324 @@ public PathsPromoteResult rollbackPathsPromote( final PathsPromoteResult result
         if ( !errors.isEmpty() )
         {
-            error = StringUtils.join( errors, "\n" );
+            List<String> rollbackErrors = promotionHelper.deleteFromStore( completed, targetStore );
+            errors.addAll( rollbackErrors );
+            result = new PathsPromoteResult( request, pending, emptySet(), emptySet(), StringUtils.join( errors, "\n" ),
+                                             validation );
         }
-
-        PathsPromoteResult result = new PathsPromoteResult( request, pending, complete, skipped, error, validation );
-
-        if ( request.isFireEvents() )
+        else
         {
-            PathsPromoteCompleteEvent evt = new PathsPromoteCompleteEvent( result );
-            fireEvent( promoteCompleteEvent, evt );
+            result = new PathsPromoteResult( request, emptySet(), completed, skipped, null, validation );
+            promotionHelper.clearStoreNFC( completed, targetStore );
+            if ( request.isFireEvents() )
+            {
+                fireEvent( promoteCompleteEvent, new PathsPromoteCompleteEvent( result ) );
+            }
         }
 
+        logger.info( "Promotion completed, promotionId: {}, timeInSeconds: {}", result.getPromotionId(),
+                     timeInSeconds( begin ) );
         return result;
     }
 
-    private static final class PathTransferResult
-    {
-        public String error;
 
-        public boolean traversed = false;
 
-        public boolean skipped = false;
+    private Callable<Set<PathTransferResult>> newPathPromotionsJob( final Collection<Transfer> transfers,
+                                                                    final ArtifactStore tgt,
+                                                                    final PathsPromoteRequest request )
+    {
+        return () -> {
+            Set<String> pathsForMDC = new HashSet<>();
+            Set<PathTransferResult> results = new HashSet<>();
+            for ( Transfer transfer : transfers )
+            {
+                pathsForMDC.add( transfer.getPath() );
+                PathTransferResult ret = doPathTransfer( transfer, tgt, request );
+                results.add( ret );
+            }
+            MDC.put( PROMOTION_CONTENT_PATH, pathsForMDC.toString() );
+            return results;
+        };
+    }
 
-        public boolean completed = false;
+    private PathTransferResult doPathTransfer( Transfer transfer, ArtifactStore tgt, PathsPromoteRequest request )
+                    throws IndyWorkflowException
+    {
+        long begin = System.currentTimeMillis();
 
-        public final String path;
+        final String path = transfer.getPath();
+        PathTransferResult result = new PathTransferResult( path );
 
-        public PathTransferResult( final String path )
+        if ( !transfer.exists() )
         {
-            this.path = path;
+            String msg = String.format( "Failed to promote: %s. Source file not exists.", transfer );
```

Comment:
We should log these too. We can't really control what happens with the promotion result, so it would be good to have these messages in our own logs so we can help debug if something is confusing.

Reply:
+1. Add info log for them.